### PR TITLE
Data viz

### PR DESCRIPTION
**Pull request by andrew032011**
### Summary

This PR is the first step towards addressing [CISLUNAR-283](https://jira.cornell.edu/browse/CISLUNAR-283) (Andrew) and [CISLUNAR-321](https://jira.cornell.edu/browse/CISLUNAR-321) (Cameron). Cameron and I pair programmed this. Currently we have plotting of the trajectory in its final position, the moon, and the earth complete.

### Testing
The outputs are all visual, so we attach a screenshot of what you see when the sim is run.
Screen Shot 2022-04-24 at 12 34 06 AM

Directions on how to run the sim and get this plot are now in the README file.

### Notes
The next step would be to get an animation/playback of this. We will put that in a separate PR.

### Comments
- [ x] Add an x between the brackets if you commented your code well!

### CODE REVIEW

**Review by sjz38 — DISMISSED** [review]
When I run `main.py` on my computer, it opens two matlab figures, one is an empty 3d graph, while the other one has the expected output. I don't know if you were having this issue but perhaps see if you are calling a figure-generating function twice. Otherwise, great stuff!

File: src/core/sim.py [hunk 1]
```diff
@@ -21,6 +23,21 @@ def __init__(self, config: Config) -> None:
...
+        self.ax.set_zlim(-10000000, 10000000)
+        self.ax.set_xlabel("x")
+        self.ax.set_ylabel("y")
+        self.ax.set_zlabel("z")
+        # self.ax = plt.axes(projection='3d')
+        self.xlocs = np.array([])
+        self.ylocs = np.array([])
+        self.zlocs = np.array([])
+
+        self.ax.set
```
[the remark this answers is not in the document]
Comment (tmf97) [reply src/core/sim.py:39]: Some ideas:
- make a plotting.py file that you can run after running a cislunarsim run that reads the data from the .csv file
- convert this entire repo over to a plugin architecture, and build a plotting plugin. I believe this is what psim does
- use ipython notebooks? (not sure about this last one)

**Review by tmf97 — CHANGES_REQUESTED** [review]
Just learned that Muhammed is working on something similar, and Sean did something similar ages ago in https://github.com/Cislunar-Explorers/FlightSoftware/blob/master/OpticalNavigation/simulations/animations.py

**Comment by andrew032011** [comment]
> When I run `main.py` on my computer, it opens two matlab figures, one is an empty 3d graph, while the other one has the expected output. I don't know if you were having this issue but perhaps see if you are calling a figure-generating function twice. Otherwise, great stuff!

I see the extra blank window. Looked into this for about an hour and dug around online. There's no "extra" matplotlib calls in my code I think. I'm not totally sure what might be causing it.

**Comment by andrew032011** [comment]
@sjz38 Modularizing the matpotlib code made the extra window glitch go away.

File: README.md [hunk 2]
```diff
@@ -10,6 +10,14 @@ source venv/bin/activate
 pip install -e .
 ```
 
+## Running the sim
+
+Make sure you're in /tests. Then run the following:
```
Comment (tmf97) [inline README.md:15]: This will get changed in #29 , FYI

File: src/core/sim.py [hunk 3]
```diff
@@ -21,6 +22,19 @@ def __init__(self, config: Config) -> None:
...
 
+        # fig = plt.figure()
+        # self.ax = fig.add_subplot(111, projection="3d")
+        # self.ax.set_xlim(-10000000, 10000000)
+        # self.ax.set_ylim(-10000000, 10000000)
+        # self.ax.set_zlim(-10000000, 10000000)
+        # self.ax.set_xlabel("x")
+        # self.ax.set_ylabel("y")
+        # self.ax.set_zlabel("z")
+        # # self.ax = plt.axes(projection='3d')
```
Comment (tmf97) [inline src/core/sim.py:33]: please delete commented code

File: src/core/sim.py [hunk 4]
```diff
@@ -47,6 +61,11 @@ def step(self) -> PropagatedOutput:
         # check if we should stop the sim
         self.should_run = not (self.should_stop())
         self.num_iters += 1
+
+        self.xlocs = np.append(self.xlocs, self.state.state.x)
+        self.ylocs = np.append(self.ylocs, self.state.state.y)
+        self.zlocs = np.append(self.zlocs, self.state.state.z)
+
```
Comment (tmf97) [inline src/core/sim.py:68]: This data is already kept track of in SimRunner's `state_history` attribute

File: src/core/sim.py [hunk 5]
```diff
@@ -72,9 +91,32 @@ def should_stop(self) -> bool:
...
+    #     self.ax.plot_surface(earth_x, earth_y, earth_z, color="g")
+
+    #     # Calculation and plotting of moon's position
+    #     moon_cx, moon_cy, moon_cz = get_body_position(self.state.time, BodyEnum.Moon)
+    #     moon_x = moon_cx + R_MOON * np.cos(u) * np.sin(v)
+    #     moon_y = moon_cy + R_MOON * np.sin(u) * np.sin(v)
+    #     moon_z = moon_cz + R_MOON * np.cos(v)
+    #     self.ax.plot_surface(moon_x, moon_y, moon_z, color="g")
+
+    #     plt.show()
```
Comment (tmf97) [inline src/core/sim.py:122]: Ditto delete

File: src/core/sim.py [hunk 6]
```diff
@@ -72,9 +91,32 @@ def should_stop(self) -> bool:
             log.error("Stopping sim because it's running too long")
             return True
 
-        if (state.x**2 + state.y**2 + state.z**2)**0.5 < R_EARTH:
+        if (state.x**2 + state.y**2 + state.z**2) ** 0.5 < R_EARTH:
             log.error("Stopping sim because craft is inside the Earth")
             log.debug(f"r={(state.x**2 + state.y**2 + state.z**2)**0.5} < {R_EARTH}")
+            data_plot = Plot(self.xlocs, self.ylocs, self.zlocs, self.state)
+            data_plot.plot_data()
```
Comment (tmf97) [inline src/core/sim.py:98]: Plotting should not occur within sim.py, only externally to keep this as modular as possible. (i.e. I'd like to be able to plot all trajectories, not just ones that intercept the earth)

File: src/utils/matplotlib_util.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        earth_y = R_EARTH * np.sin(u) * np.sin(v)
+        earth_z = R_EARTH * np.cos(v)
+        self.ax.plot_surface(earth_x, earth_y, earth_z, color="g")
+
+        # Calculation and plotting of moon's position
+        moon_cx, moon_cy, moon_cz = get_body_position(self.state.time, BodyEnum.Moon)
+        moon_x = moon_cx + R_MOON * np.cos(u) * np.sin(v)
+        moon_y = moon_cy + R_MOON * np.sin(u) * np.sin(v)
+        moon_z = moon_cz + R_MOON * np.cos(v)
+        self.ax.plot_surface(moon_x, moon_y, moon_z, color="gray")
```
Comment (tmf97) [inline src/utils/matplotlib_util.py:40]: The moon's position changes throughout a sim, but this only plots it at one point in time? It's OK if we're only plotting the last position, but def mention that in a comment somewhere.

File: src/utils/matplotlib_util.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+        # Calculation and plotting of sun's position
+        sun_cx, sun_cy, sun_cz = get_body_position(self.state.time, BodyEnum.Sun)
+        sun_x = sun_cx + R_SUN * np.cos(u) * np.sin(v)
+        sun_y = sun_cy + R_SUN * np.sin(u) * np.sin(v)
+        sun_z = sun_cz + R_SUN * np.cos(v)
+        self.ax.plot_surface(sun_x, sun_y, sun_z, color="y")
```
Comment (tmf97) [inline src/utils/matplotlib_util.py:47]: The sun is SUUUUUUUUUUUPER far away and we generally don't care about plotting it. Don't delete this, but don't leave it as a default

**Review by tmf97 — CHANGES_REQUESTED** [review]

**Review by tmf97 — COMMENTED** [review]
I can't get matplotlib to show the plots, but this does look good.
Just merge main into this and we should be good.

**Comment by andrew032011** [comment]
> I can't get matplotlib to show the plots on my machine so i can't double check these plots, but this code does look good. Just merge main into this and we should be good.

I updated the screenshots of what we get now. It's two plots, one 2D and one 3D, next to each other. We can then add more plots in the future based on what exactly we want to show.

**Review by tmf97 — APPROVED** [review]
I think this is in a solid spot - we can always add more functionality later.
I tried using `plt.savefig` to checkout the results on my machine, but it didn't plot the Earth/moon surfaces - no big deal though.

File: src/utils/plot.py [hunk 9]
```diff
@@ -0,0 +1,28 @@
+import pandas as pd
+from utils.matplotlib_util import Plot
+
+
+import argparse
+
+
+class PlotHelper:
```
Comment (sjz38) [inline src/utils/plot.py:8]: How do you run this? I tried running `python3 src/utils/matplotlib_util.py runs/cislunarsim-1650564840.722935.csv` but I don't see a matplotlib window open. Is this the right command? Please comment how to run this in the readme

Comment (andrew032011) [reply src/utils/plot.py:8]: In the readme, it states that the command is
python3 src/utils/plot.py {file path}

Comment (sjz38) [reply src/utils/plot.py:8]: Oops forgot to pull all good 👍

**Review by sjz38 — APPROVED** [review]
👍

**Pull request merged**